controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,14 +134,335 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
+
+			// storageClassClaimStorageType is the storage type of the StorageClassClaim
+			var storageClassClaimStorageType providerpb.FulfillStorageClassClaimRequest_StorageType
+			switch r.storageClassClaim.Spec.Type {
+			case "blockpool":
+				storageClassClaimStorageType = 1
+			case "sharedfilesystem":
+				storageClassClaimStorageType = 2
+			}
+
+			// Call the `FulfillStorageClass` service on the provider server with StorageClassClaim as a request message.
+			_, err = providerClient.FulfillStorageClassClaim(r.ctx, r.storageCluster.Status.ExternalStorage.ConsumerID, r.storageClassClaim.Name, r.storageClassClaim.Spec.EncryptionMethod, storageClassClaimStorageType)
+			if err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to fulfill StorageClassClaim: %v", err)
+			}
+
+			// Update the StorageClassClaim status.
+			r.storageClassClaim.Status.Phase = ocsv1alpha1.StorageClassClaimCreating
+
+			// Call the `GetStorageClassClaimConfig` service on the provider server with StorageClassClaim as a request message.
+			response, err := providerClient.GetStorageClassClaimConfig(r.ctx, r.storageCluster.Status.ExternalStorage.ConsumerID, r.storageClassClaim.Name)
+			if err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to get StorageClassClaim config: %v", err)
+			}
+
+			// Populate the externalResources with the formatted response data.
+			var externalResources []ExternalResource
+			resources := response.ExternalResource
+			if resources == nil {
+				return reconcile.Result{}, fmt.Errorf("no configuration data recieved")
+			}
+
+			// Unmarshal the response datum as externalResource.
+			for _, resource := range resources {
+				data := map[string]string{}
+				err = json.Unmarshal(resource.Data, &data)
+				if err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to unmarshal StorageClassClaim configuration response: %v", err)
+				}
+				externalResources = append(externalResources, ExternalResource{
+					Name: resource.Name,
+					Data: data,
+					Kind: resource.Kind,
+				})
+			}
+
+			// Go over the received objects and operate on them accordingly.
+			var enableRookCSICephFS bool
+			var availableStorageClassConfigurations []StorageClassConfiguration
+			var storageClassClaimOwnerReference metav1.OwnerReference
+			for _, resource := range externalResources {
+
+				// storageClassClaimOwnerReference points to the StorageClassClaim object as the owner object of the created resource(s).
+				storageClassClaimOwnerReference = metav1.OwnerReference{
+					APIVersion: r.storageClassClaim.APIVersion,
+					Kind:       r.storageClassClaim.Kind,
+					Name:       r.storageClassClaim.Name,
+					UID:        r.storageClassClaim.UID,
+				}
+				switch resource.Kind {
+				case "Secret":
+					objectMeta := metav1.ObjectMeta{
+						Name:      resource.Name,
+						Namespace: r.storageClassClaim.Namespace,
+						OwnerReferences: []metav1.OwnerReference{
+							storageClassClaimOwnerReference,
+						},
```

Comment:
> Do we want ctrl.SetControllerReference or ctrl.SetOwnerReference?
Interesting question, maybe SetOwnerReference is better in this case.

> I don't see the value add of r.Own utility.
Same reasoning as the one I gave in the comment about all of the other suggested utils

Reply:
Only major difference I can notice between the two is (w.r.t. `SetControllerReference`):

> This is used for **garbage collection** of the controlled object and for **reconciling the owner object on changes to controlled** (with a Watch + EnqueueRequestForOwner).

> Since **only one OwnerReference can be a controller**, it returns an error if there is another OwnerReference with Controller flag set.

If the above if is not needed (which I think we do), then we can just replace all instances with `SetOwnerReference`.